File: csharp/src/Warehouse.Adapters/Ids.cs
using Warehouse.Application;
using Warehouse.Domain;

namespace Warehouse.Adapters;

public sealed class SequenceOrderIdGenerator : IOrderIdGenerator
{
    private readonly string _prefix;
    private readonly object _gate = new();
    private int _n;

    public SequenceOrderIdGenerator(string prefix = "ord")
    {
        ArgumentNullException.ThrowIfNull(prefix);
        _prefix = prefix;
    }

    public OrderId Next()
    {
        lock (_gate)
        {
            _n++;
            return new OrderId($"{_prefix}-{_n}");
        }
    }
}

public sealed class FixedOrderIdGenerator : IOrderIdGenerator
{
    private readonly OrderId _orderId;

    public FixedOrderIdGenerator(OrderId orderId)
    {
        _orderId = orderId;
    }

    public OrderId Next() => _orderId;
}

File: swift/Sources/Warehouse/Adapters/Ids.swift
import Dispatch

/// Test double that always returns the same injected identifier.
public struct FixedOrderIdGenerator: OrderIdGenerator {
  private let orderId: OrderId

  public init(_ orderId: OrderId) {
    self.orderId = orderId
  }

  public func next() -> OrderId {
    orderId
  }
}

/// Test double that issues `ord-1`, `ord-2`, ...
public final class SequenceOrderIdGenerator: OrderIdGenerator, @unchecked Sendable {
  private let queue = DispatchQueue(label: "warehouse.ids")
  private var n = 0
  private let prefix: String

  public init(prefix: String = "ord") {
    self.prefix = prefix
  }

  public func next() -> OrderId {
    queue.sync {
      n += 1
      return OrderId(sequence: n, prefix: prefix)
    }
  }
}
